src/v/test_utils/randoms.h

```diff
@@ -39,4 +46,104 @@ random_broker(int32_t id_low_bound, int32_t id_upper_bound) {
 }
 
-} // namespace tests
\ No newline at end of file
+inline bool random_bool() {
+    static thread_local std::random_device
+      rd; // Will be used to obtain a seed for the random number engine
+    static thread_local std::mt19937 gen(
+      rd()); // Standard mersenne_twister_engine seeded with rd()
+    std::uniform_int_distribution<> dist(0, 100);
+    return dist(gen) > 50;
+}
+
+template<typename T>
+const T& random_choice(const std::vector<T>& elements) {
+    auto idx = random_generators::get_int<size_t>(0, elements.size() - 1);
+    return elements[idx];
+}
+
+template<typename T>
+T random_named_string(size_t size = 20) {
+    return T{random_generators::gen_alphanum_string(size)};
+}
+
+template<typename T>
+T random_named_int() {
+    return T{random_generators::get_int<typename T::type>(
+      0, std::numeric_limits<T>::max())};
+}
+
+inline model::topic_namespace random_topic_namespace() {
+    return {
+      random_named_string<model::ns>(), random_named_string<model::topic>()};
+}
+
+template<typename Func>
+auto random_optional(Func f) {
+    using T = decltype(f());
+    if (random_bool()) {
+        return std::optional<T>(f());
+    }
+    return std::optional<T>();
+}
+
+template<typename Func>
+auto random_tristate(Func f) {
+    using T = decltype(f());
+    if (random_bool()) {
+        return tristate<T>{};
+    }
+    return tristate<T>(random_optional(f));
+}
+
+inline model::ntp random_ntp() {
+    return {
+      random_named_string<model::ns>(),
+      random_named_string<model::topic>(),
+      random_named_int<model::partition_id>()};
+}
+
+inline model::cleanup_policy_bitflags random_cleanup_policy() {
+    return random_choice(std::vector<model::cleanup_policy_bitflags>{
+      model::cleanup_policy_bitflags::none,
+      model::cleanup_policy_bitflags::compaction,
+      model::cleanup_policy_bitflags::deletion});
+}
+
+inline model::compaction_strategy random_compaction_strategy() {
+    return tests::random_choice(std::vector<model::compaction_strategy>{
+      model::compaction_strategy::header,
+      model::compaction_strategy::offset,
+      model::compaction_strategy::timestamp});
+}
+
+inline model::compression random_compression() {
+    return random_choice(std::vector<model::compression>{
+      model::compression::gzip,
+      model::compression::zstd,
+      model::compression::none,
+      model::compression::producer,
+      model::compression::snappy});
+}
+
+inline model::shadow_indexing_mode random_shadow_indexing_mode() {
+    return tests::random_choice(std::vector<model::shadow_indexing_mode>{
+      model::shadow_indexing_mode::archival,
+      model::shadow_indexing_mode::drop_archival,
+      model::shadow_indexing_mode::drop_fetch,
+      model::shadow_indexing_mode::drop_full,
+      model::shadow_indexing_mode::disabled,
+      model::shadow_indexing_mode::full,
+    });
+}
+
+inline model::timestamp_type random_timestamp_type() {
```

Comment:
i think these should be in model test utilities. the moment one of these types isn't header-only we start creating awkward dependencies.